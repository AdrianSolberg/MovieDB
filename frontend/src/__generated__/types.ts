export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Date: { input: any; output: any; }
};

export type Movie = {
  _id: Scalars['Int']['output'];
  title: Scalars['String']['output'];
  vote_average: Scalars['Float']['output'];
  vote_count: Scalars['Int']['output'];
  status: Scalars['String']['output'];
  release_date: Scalars['Date']['output'];
  revenue: Scalars['Float']['output'];
  runtime: Scalars['Int']['output'];
  backdrop_path?: Maybe<Scalars['String']['output']>;
  budget: Scalars['Int']['output'];
  homepage?: Maybe<Scalars['String']['output']>;
  imdb_id?: Maybe<Scalars['String']['output']>;
  original_language: Scalars['String']['output'];
  original_title: Scalars['String']['output'];
  overview: Scalars['String']['output'];
  popularity: Scalars['Float']['output'];
  poster_path?: Maybe<Scalars['String']['output']>;
  tagline?: Maybe<Scalars['String']['output']>;
  genres: Array<Scalars['String']['output']>;
  production_companies: Array<Scalars['String']['output']>;
  production_countries: Array<Scalars['String']['output']>;
  spoken_languages: Array<Scalars['String']['output']>;
  keywords: Array<Scalars['String']['output']>;
};

export type Filters = {
  Genre: Array<Scalars['String']['output']>;
  Rating: Array<Scalars['String']['output']>;
  Decade: Array<Scalars['String']['output']>;
  Status: Array<Scalars['String']['output']>;
  Runtime: Array<Scalars['String']['output']>;
};

export type FiltersInput = {
  Genre: Array<Scalars['String']['input']>;
  Rating: Array<Scalars['String']['input']>;
  Decade: Array<Scalars['String']['input']>;
  Status: Array<Scalars['String']['input']>;
  Runtime: Array<Scalars['String']['input']>;
};

export enum SortingType {
  NewestFirst = 'NEWEST_FIRST',
  OldestFirst = 'OLDEST_FIRST',
  BestRated = 'BEST_RATED',
  WorstRated = 'WORST_RATED',
  LongestRuntime = 'LONGEST_RUNTIME',
  ShortestRuntime = 'SHORTEST_RUNTIME'
}

export type Query = {
  movie?: Maybe<Movie>;
  movies: Array<Movie>;
  filters?: Maybe<Filters>;
};


export type QueryMovieArgs = {
  id: Scalars['Int']['input'];
};


export type QueryMoviesArgs = {
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
};

export type GetMovieQueryVariables = Exact<{
  movieId: Scalars['Int']['input'];
}>;


export type GetMovieQuery = { __typename?: 'Query', movie?: { __typename?: 'Movie', _id: number, title: string, vote_average: number, vote_count: number, status: string, release_date: any, revenue: number, runtime: number, backdrop_path?: string | null, budget: number, homepage?: string | null, imdb_id?: string | null, original_language: string, original_title: string, overview: string, popularity: number, poster_path?: string | null, tagline?: string | null, genres: Array<string>, production_companies: Array<string>, production_countries: Array<string>, spoken_languages: Array<string>, keywords: Array<string> } | null };

export type GetMoviesQueryVariables = Exact<{
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
}>;


export type GetMoviesQuery = { __typename?: 'Query', movies: Array<{ __typename?: 'Movie', _id: number, title: string, vote_average: number, release_date: any, runtime: number, poster_path?: string | null }> };

export type GetFiltersQueryVariables = Exact<{ [key: string]: never; }>;


export type GetFiltersQuery = { __typename?: 'Query', filters?: { __typename?: 'Filters', Genre: Array<string>, Rating: Array<string>, Decade: Array<string>, Status: Array<string>, Runtime: Array<string> } | null };
