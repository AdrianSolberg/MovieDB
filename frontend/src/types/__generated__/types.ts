export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = {
    [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]: Maybe<T[SubKey]>;
};
export type MakeEmpty<
    T extends { [key: string]: unknown },
    K extends keyof T,
> = { [_ in K]?: never };
export type Incremental<T> =
    | T
    | {
          [P in keyof T]?: P extends " $fragmentName" | "__typename"
              ? T[P]
              : never;
      };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
    ID: { input: string; output: string };
    String: { input: string; output: string };
    Boolean: { input: boolean; output: boolean };
    Int: { input: number; output: number };
    Float: { input: number; output: number };
    DateTime: { input: Date; output: Date };
};

export type Movie = {
    __typename?: "Movie";
    _id: Scalars["Int"]["output"];
    title: Scalars["String"]["output"];
    vote_average: Scalars["Float"]["output"];
    vote_count: Scalars["Int"]["output"];
    status: Scalars["String"]["output"];
    release_date: Scalars["DateTime"]["output"];
    revenue: Scalars["Float"]["output"];
    runtime: Scalars["Int"]["output"];
    backdrop_path?: Maybe<Scalars["String"]["output"]>;
    budget: Scalars["Int"]["output"];
    homepage?: Maybe<Scalars["String"]["output"]>;
    imdb_id?: Maybe<Scalars["String"]["output"]>;
    original_language: Scalars["String"]["output"];
    original_title: Scalars["String"]["output"];
    overview: Scalars["String"]["output"];
    popularity: Scalars["Float"]["output"];
    poster_path?: Maybe<Scalars["String"]["output"]>;
    tagline?: Maybe<Scalars["String"]["output"]>;
    genres: Array<Scalars["String"]["output"]>;
    production_companies: Array<Scalars["String"]["output"]>;
    production_countries: Array<Scalars["String"]["output"]>;
    spoken_languages: Array<Scalars["String"]["output"]>;
    keywords: Array<Scalars["String"]["output"]>;
    reviews: Array<Review>;
};

export type Review = {
    __typename?: "Review";
    _id: Scalars["ID"]["output"];
    movie: Movie;
    username: Scalars["String"]["output"];
    rating: Scalars["Int"]["output"];
    comment: Scalars["String"]["output"];
    date: Scalars["DateTime"]["output"];
};

export type Filters = {
    __typename?: "Filters";
    Genre: Array<Scalars["String"]["output"]>;
    Rating: Array<Scalars["String"]["output"]>;
    Decade: Array<Scalars["String"]["output"]>;
    Status: Array<Scalars["String"]["output"]>;
    Runtime: Array<Scalars["String"]["output"]>;
};

export type FiltersInput = {
    Genre: Array<Scalars["String"]["input"]>;
    Rating: Array<Scalars["String"]["input"]>;
    Decade: Array<Scalars["String"]["input"]>;
    Status: Array<Scalars["String"]["input"]>;
    Runtime: Array<Scalars["String"]["input"]>;
};

export enum SortingType {
    MOST_POPULAR = "MOST_POPULAR",
    NEWEST_FIRST = "NEWEST_FIRST",
    OLDEST_FIRST = "OLDEST_FIRST",
    BEST_RATED = "BEST_RATED",
    WORST_RATED = "WORST_RATED",
    LONGEST_RUNTIME = "LONGEST_RUNTIME",
    SHORTEST_RUNTIME = "SHORTEST_RUNTIME",
}

export type Query = {
    __typename?: "Query";
    movie?: Maybe<Movie>;
    movies: Array<Movie>;
    filters: Filters;
    latestReviews: Array<Review>;
    userReviews: Array<Review>;
};

export type QuerymovieArgs = {
    id: Scalars["Int"]["input"];
};

export type QuerymoviesArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
};

export type QuerylatestReviewsArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type QueryuserReviewsArgs = {
    username: Scalars["String"]["input"];
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type Mutation = {
    __typename?: "Mutation";
    addReview: Movie;
    deleteReview: Movie;
};

export type MutationaddReviewArgs = {
    movie_id: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
};

export type MutationdeleteReviewArgs = {
    _id: Scalars["ID"]["input"];
};

export type GetMoviesQueryVariables = Exact<{
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
}>;

export type GetMoviesQuery = {
    __typename?: "Query";
    movies: Array<{
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        release_date: Date;
        runtime: number;
        poster_path?: string | null;
    }>;
};

export type GetMovieQueryVariables = Exact<{
    movieId: Scalars["Int"]["input"];
}>;

export type GetMovieQuery = {
    __typename?: "Query";
    movie?: {
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        vote_count: number;
        status: string;
        release_date: Date;
        revenue: number;
        runtime: number;
        backdrop_path?: string | null;
        budget: number;
        homepage?: string | null;
        imdb_id?: string | null;
        original_language: string;
        original_title: string;
        overview: string;
        popularity: number;
        poster_path?: string | null;
        tagline?: string | null;
        genres: Array<string>;
        production_companies: Array<string>;
        production_countries: Array<string>;
        spoken_languages: Array<string>;
        keywords: Array<string>;
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    } | null;
};

export type GetFiltersQueryVariables = Exact<{ [key: string]: never }>;

export type GetFiltersQuery = {
    __typename?: "Query";
    filters: {
        __typename?: "Filters";
        Genre: Array<string>;
        Rating: Array<string>;
        Decade: Array<string>;
        Status: Array<string>;
        Runtime: Array<string>;
    };
};

export type AddReviewMutationVariables = Exact<{
    movieId: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
}>;

export type AddReviewMutation = {
    __typename?: "Mutation";
    addReview: {
        __typename?: "Movie";
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    };
};
