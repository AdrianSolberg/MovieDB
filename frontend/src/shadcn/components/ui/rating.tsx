"use client";

import React from "react";
import { Star } from "lucide-react";

import { cn } from "../../lib/utils";

const ratingVariants = {
    default: {
        star: "text-foreground",
        emptyStar: "text-muted-foreground",
    },
    destructive: {
        star: "text-destructive",
        emptyStar: "text-destructive/70",
    },
    yellow: {
        star: "text-yellow-500",
        emptyStar: "text-yellow-200",
    },
};

interface RatingsProps extends React.HTMLAttributes<HTMLDivElement> {
    totalstars?: number;
    size?: number;
    fill?: boolean;
    Icon?: React.ReactElement;
    variant?: keyof typeof ratingVariants;
    asInput?: boolean;
    value: number;
    onValueChange?: (value: number) => void;
}

const Ratings = ({ ...props }: RatingsProps) => {
    const {
        totalstars = 5,
        size = 20,
        fill = true,
        Icon = <Star />,
        variant = "default",
        asInput = false,
        onValueChange,
        value,
        ...extraProps
    } = props;

    const handleKeyDown = (e: React.KeyboardEvent, index: number) => {
        if (asInput && e.key === " ") {
            // " " is the encoding for spacebar
            e.preventDefault();
            onValueChange && onValueChange(index + 1);
        }
    };

    const ratings = value;

    const fullStars = Math.floor(ratings);
    const partialStar =
        ratings % 1 > 0 ? (
            <PartialStar
                fillPercentage={ratings % 1}
                size={size}
                className={cn(ratingVariants[variant].star)}
                Icon={Icon}
                asInput={asInput}
                onValueChange={() =>
                    onValueChange && onValueChange(fullStars + 1)
                }
            />
        ) : null;

    return (
        <div
            className={cn("flex items-center gap-2")}
            role={asInput ? "radiogroup" : undefined}
            {...extraProps}
        >
            {[...Array(fullStars)].map((_, i) =>
                React.cloneElement(Icon, {
                    key: i,
                    size,
                    className: cn(
                        fill ? "fill-current" : "fill-transparent",
                        ratingVariants[variant].star,
                        asInput ? "cursor-pointer" : ""
                    ),
                    role: asInput ? "radio" : undefined,
                    "aria-checked": asInput ? i + 1 === value : undefined,
                    "aria-label": asInput ? `${i + 1} star` : undefined,
                    tabIndex: asInput ? 0 : -1,
                    onClick: () => onValueChange && onValueChange(i + 1),
                    onKeyDown: asInput
                        ? (e: React.KeyboardEvent) => handleKeyDown(e, i)
                        : undefined,
                })
            )}
            {partialStar}
            {[...Array(totalstars - fullStars - (partialStar ? 1 : 0))].map(
                (_, i) =>
                    React.cloneElement(Icon, {
                        key: i + fullStars + 1,
                        size,
                        className: cn(
                            ratingVariants[variant].emptyStar,
                            asInput ? "cursor-pointer" : ""
                        ),
                        role: asInput ? "radio" : undefined,
                        "aria-checked": asInput
                            ? fullStars + i + 1 + (partialStar ? 1 : 0) ===
                              value
                            : undefined,
                        tabIndex: asInput ? 0 : -1,
                        onKeyDown: asInput
                            ? (e: React.KeyboardEvent) =>
                                  handleKeyDown(
                                      e,
                                      fullStars + i + (partialStar ? 1 : 0)
                                  )
                            : undefined,
                        "aria-label": asInput
                            ? `${fullStars + i + 1 + (partialStar ? 1 : 0)} star`
                            : undefined,
                        onClick: () =>
                            onValueChange &&
                            onValueChange(
                                fullStars + i + 1 + (partialStar ? 1 : 0)
                            ),
                    })
            )}
        </div>
    );
};

interface PartialStarProps {
    fillPercentage: number;
    size: number;
    className?: string;
    Icon: React.ReactElement;
    asInput?: boolean;
    onValueChange?: () => void;
}

const PartialStar = ({ ...props }: PartialStarProps) => {
    const { fillPercentage, size, className, Icon, asInput, onValueChange } =
        props;

    return (
        <div
            role={asInput ? "input" : undefined}
            onClick={() => onValueChange && onValueChange()}
            className={cn("relative inline-block", asInput && "cursor-pointer")}
        >
            {React.cloneElement(Icon, {
                size,
                className: cn("fill-transparent", className),
            })}
            <div
                style={{
                    position: "absolute",
                    top: 0,
                    overflow: "hidden",
                    width: `${fillPercentage * 100}%`,
                }}
            >
                {React.cloneElement(Icon, {
                    size,
                    className: cn("fill-current", className),
                })}
            </div>
        </div>
    );
};

export default Ratings;
