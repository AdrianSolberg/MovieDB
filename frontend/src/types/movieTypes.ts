/* eslint-disable @typescript-eslint/no-empty-object-type */

// We disable the eslint rule above because we want to use empty interfaces to
// differentiate between different types even though they have the same structure

interface IdName {
    id: number;
    name: string;
}

interface Genre extends IdName {}
interface Keyword extends IdName {}
interface Language extends IdName {}
interface ProductionCompany extends IdName {}
interface ProductionCountry extends IdName {}

enum Status {
    Released = "Released",
    InProduction = "In Production",
    PostProduction = "Post Production",
    Planned = "Planned",
    Rumored = "Rumored",
    Canceled = "Canceled",
}

interface Movie {
    id: number; // Unique identifier
    title: string; // Unique in combination with release_date
    vote_average: number;
    vote_count: number;
    status: Status;
    release_date: Date;
    revenue: number;
    runtime: number;
    adult: boolean;
    backdrop_path: string | undefined;
    budget: number;
    homepage: string | undefined;
    imdb_id: string | undefined;
    original_language: string;
    original_title: string;
    overview: string;
    popularity: number;
    poster_path: string | undefined;
    tagline: string | undefined;
    genres: Genre[];
    production_companies: ProductionCompany[];
    production_countries: ProductionCountry[];
    spoken_languages: Language[];
    keywords: Keyword[];
}

export { Status };
export type {
    Genre,
    Keyword,
    Language,
    ProductionCompany,
    ProductionCountry,
    Movie,
};
