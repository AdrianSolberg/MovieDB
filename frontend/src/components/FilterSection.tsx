import { Filter, FiltersInput } from "@/types/__generated__/types";
import {
    AccordionContent,
    AccordionItem,
    AccordionTrigger,
} from "../shadcn/components/ui/accordion";
import { Checkbox } from "../shadcn/components/ui/checkbox";
import { formatNumber } from "@/utils/formatUtil";

interface FilterSectionInterface {
    category: keyof FiltersInput;
    all_filters: Filter[];
    applied_filters: string[];
    updateFilters: (category: keyof FiltersInput, filter: string) => void;
    loading: boolean;
}

const FilterSection: React.FC<FilterSectionInterface> = ({
    category,
    all_filters,
    applied_filters,
    updateFilters,
    loading,
}) => {
    return (
        <AccordionItem value={`${category} item`}>
            <AccordionTrigger>
                <p>
                    {category}
                    {applied_filters.length > 0 && (
                        <span className="opacity-60">
                            {" "}
                            ({applied_filters.length} applied)
                        </span>
                    )}
                </p>
            </AccordionTrigger>
            <AccordionContent>
                <ul>
                    {all_filters.map(
                        (filter) =>
                            filter.hits > 0 && (
                                <li
                                    key={filter.name}
                                    className="mb-2 flex items-center space-x-2"
                                >
                                    <Checkbox
                                        id={filter.name}
                                        checked={applied_filters.includes(
                                            filter.name
                                        )}
                                        disabled={loading}
                                        onCheckedChange={() =>
                                            updateFilters(category, filter.name)
                                        }
                                    />
                                    <label
                                        htmlFor={filter.name}
                                        className="text-sm font-medium leading-none hover:cursor-pointer"
                                    >
                                        {filter.name}{" "}
                                        <span className="opacity-60">
                                            ({formatNumber(filter.hits)})
                                        </span>
                                    </label>
                                </li>
                            )
                    )}
                </ul>
            </AccordionContent>
        </AccordionItem>
    );
};

export default FilterSection;
