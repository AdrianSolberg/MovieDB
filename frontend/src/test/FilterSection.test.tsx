import { vi } from "vitest";
import { render, screen } from "@testing-library/react";
import FilterSection from "@/components/FilterSection";
import { FiltersInput } from "@/types/__generated__/types";
import { Accordion } from "@/shadcn/components/ui/accordion";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

vi.mock("@/utils/formatUtil", () => ({
    formatNumber: (n: number) => n.toLocaleString("no-NO"),
}));

const mockUpdateFilters = vi.fn();

const defaultProps = {
    category: "Genre" as keyof FiltersInput,
    all_filters: [
        { name: "Action", hits: 150 },
        { name: "Comedy", hits: 100 },
        { name: "Drama", hits: 120 },
        { name: "Sciende Fiction", hits: 0 },
        { name: "Thriller", hits: 90 },
    ],
    applied_filters: ["Action"],
    updateFilters: mockUpdateFilters,
    loading: false,
};

describe("FilterSection", () => {
    const renderFilterSection = async (props: {
        category: keyof FiltersInput;
        all_filters: { name: string; hits: number }[];
        applied_filters: string[];
        updateFilters: (category: keyof FiltersInput, filter: string) => void;
        loading: boolean;
    }) => {
        render(
            <Accordion type="single" collapsible className="w-full">
                <FilterSection {...props} />
            </Accordion>
        );

        await userEvent.click(screen.getByRole("button"));
    };

    it("renders the category name correctly", () => {
        renderFilterSection(defaultProps);
        expect(screen.getByText("Genre")).toBeInTheDocument();
    });

    it("shows the number of applied filters", () => {
        renderFilterSection(defaultProps);
        expect(screen.getByText("(1 applied)")).toBeInTheDocument();
    });

    it("does not show applied filters count when none are applied", () => {
        renderFilterSection({ ...defaultProps, applied_filters: [] });
        expect(screen.queryByText("(0 applied)")).not.toBeInTheDocument();
    });

    it("only renders filters with hits greater than 0", async () => {
        await renderFilterSection(defaultProps);

        expect(screen.getByText("Action")).toBeInTheDocument();
        expect(screen.getByText("Comedy")).toBeInTheDocument();
        expect(screen.getByText("Drama")).toBeInTheDocument();
        expect(screen.getByText("Thriller")).toBeInTheDocument();
        expect(screen.queryByText("Science Fiction")).not.toBeInTheDocument();
    });

    it("shows the correct number of hits for each filter", async () => {
        await renderFilterSection(defaultProps);

        expect(screen.getByText("(150)")).toBeInTheDocument();
        expect(screen.getByText("(100)")).toBeInTheDocument();
        expect(screen.getByText("(120)")).toBeInTheDocument();
        expect(screen.getByText("(90)")).toBeInTheDocument();
    });

    it("checks the checkbox only for applied filters", async () => {
        await renderFilterSection(defaultProps);

        const checkboxes = screen.getAllByRole("checkbox");
        checkboxes.forEach((checkbox) => {
            if (checkbox.id === "Action") {
                expect(checkbox).toBeChecked();
                return;
            }
            expect(checkbox).not.toBeChecked();
        });
    });

    it("calls updateFilters with correct arguments when checkbox is clicked", async () => {
        await renderFilterSection(defaultProps);

        const comedyCheckbox = screen.getByRole("checkbox", {
            name: /Comedy/,
        });
        await userEvent.click(comedyCheckbox);
        expect(mockUpdateFilters).toHaveBeenCalledWith("Genre", "Comedy");
    });

    it("disables checkboxes when loading is true", async () => {
        await renderFilterSection({ ...defaultProps, loading: true });

        const checkboxes = screen.getAllByRole("checkbox");
        checkboxes.forEach((checkbox) => {
            expect(checkbox).toBeDisabled();
        });
    });

    it("enables checkboxes when loading is false", async () => {
        await renderFilterSection(defaultProps);

        const checkboxes = screen.getAllByRole("checkbox");
        checkboxes.forEach((checkbox) => {
            expect(checkbox).not.toBeDisabled();
        });
    });
});
