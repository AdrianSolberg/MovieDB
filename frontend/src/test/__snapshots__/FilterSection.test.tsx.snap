// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterSection > matches snapshot when collapsed 1`] = `
<DocumentFragment>
  <div
    class="w-full"
    data-orientation="vertical"
  >
    <div
      class="border-b"
      data-orientation="vertical"
      data-state="closed"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="closed"
      >
        <button
          aria-controls="radix-:r1:"
          aria-expanded="false"
          class="flex flex-1 items-center justify-between py-4 font-medium transition-all hover:underline [&[data-state=open]>svg]:rotate-180"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="closed"
          id="radix-:r0:"
          type="button"
        >
          <p>
            Genre
            <span
              class="opacity-60"
            >
               (1 applied)
            </span>
          </p>
          <svg
            class="lucide lucide-chevron-down h-4 w-4 shrink-0 transition-transform duration-200"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </h3>
      <div
        aria-labelledby="radix-:r0:"
        class="max-h-40 overflow-y-auto text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
        data-orientation="vertical"
        data-state="closed"
        hidden=""
        id="radix-:r1:"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`FilterSection > matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="w-full"
    data-orientation="vertical"
  >
    <div
      class="border-b"
      data-orientation="vertical"
      data-state="open"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="open"
      >
        <button
          aria-controls="radix-:r3:"
          aria-expanded="true"
          class="flex flex-1 items-center justify-between py-4 font-medium transition-all hover:underline [&[data-state=open]>svg]:rotate-180"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="open"
          id="radix-:r2:"
          type="button"
        >
          <p>
            Genre
            <span
              class="opacity-60"
            >
               (1 applied)
            </span>
          </p>
          <svg
            class="lucide lucide-chevron-down h-4 w-4 shrink-0 transition-transform duration-200"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </h3>
      <div
        aria-labelledby="radix-:r2:"
        class="max-h-40 overflow-y-auto text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
        data-orientation="vertical"
        data-state="open"
        id="radix-:r3:"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      >
        <div
          class="pb-4 pt-0"
        >
          <ul>
            <li
              class="mb-2 flex items-center space-x-2"
            >
              <button
                aria-checked="true"
                aria-label="Action"
                class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
                data-state="checked"
                id="Action"
                role="checkbox"
                type="button"
                value="on"
              >
                <span
                  class="flex items-center justify-center text-current"
                  data-state="checked"
                  style="pointer-events: none;"
                >
                  <svg
                    class="lucide lucide-check h-4 w-4"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M20 6 9 17l-5-5"
                    />
                  </svg>
                </span>
              </button>
              <label
                class="text-sm font-medium leading-none hover:cursor-pointer"
                for="Action"
              >
                Action 
                <span
                  class="opacity-60"
                >
                  (150)
                </span>
              </label>
            </li>
            <li
              class="mb-2 flex items-center space-x-2"
            >
              <button
                aria-checked="false"
                aria-label="Comedy"
                class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
                data-state="unchecked"
                id="Comedy"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="text-sm font-medium leading-none hover:cursor-pointer"
                for="Comedy"
              >
                Comedy 
                <span
                  class="opacity-60"
                >
                  (100)
                </span>
              </label>
            </li>
            <li
              class="mb-2 flex items-center space-x-2"
            >
              <button
                aria-checked="false"
                aria-label="Drama"
                class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
                data-state="unchecked"
                id="Drama"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="text-sm font-medium leading-none hover:cursor-pointer"
                for="Drama"
              >
                Drama 
                <span
                  class="opacity-60"
                >
                  (120)
                </span>
              </label>
            </li>
            <li
              class="mb-2 flex items-center space-x-2"
            >
              <button
                aria-checked="false"
                aria-label="Thriller"
                class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
                data-state="unchecked"
                id="Thriller"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="text-sm font-medium leading-none hover:cursor-pointer"
                for="Thriller"
              >
                Thriller 
                <span
                  class="opacity-60"
                >
                  (90)
                </span>
              </label>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
